Documentation for the mixer (#146)
ExtremeFLOW/neko-top: documentation/pages/tutorials/mixer.md

# Static Mixer {#mixer}

\tableofcontents

## Optimization problem

This tutorial aims to replicate the passive mixer problem by 
[C. S. Andreasen et al. 2009](https://doi.org/10.1002/fld.1964).

The goal of this optimization problem is to design a structure capable of

> mixing fluids in a static mannor, that is to say, without the use of moving

Reviewer: ```suggestion
mixing fluids in a static manor, that is to say, without the use of moving
```

(denoted in green in the above figure). The
system is assumed to be low Reynolds number, implying a steady state solution
can be found, and high Peclet number, implying the mixing must be performed
by advective transport, not diffusion.

In this tutorial you will learn how to
- [Formally define this problem in the context of a topology optimization problem.](@ref mixer_optimization_problem)
- [Mesh the domain.](@ref mixer_meshing)
- [Set up a fluid, scalar and their respective adjoint simulations.](@ref mixer_case_file)
- [Define different domains](@ref mixer_domains)
- [Define the objective functions and constraints being solved.](@ref mixer_objectives)
- [Define a mapping cascade.](@ref mixer_mapping)
- [Define optimization parameters for the MMA algorithm.](@ref mixer_MMA)
- [Prepare a driver to solve the optimization problem.](@ref mixer_driver)
- [Solve the optimization problem.](@ref mixer_solve)
- [Post process the results.](@ref mixer_post)

## Defining the optimization problem {#mixer_optimization_problem}

![Indication of the computational domain and boundary conditions.](mixer_BCs.png)

We define the duct as the domain \f$\Omega\f$.

> The equations being solved in \f$\Omega\f$ are the incompressible Navier--Stokes

Reviewer: ```suggestion
The equations being solved in \f$\Omega\f$ are the incompressible Navier-Stokes
```

`neko-top` provides users with the capability of executing custom scripts before
an the simulation begins but including a `prepare.sh` script in the example 
folder. This functionality is commonly used to generate
meshes or other preprocessing steps.

An extensive review of the meshing tools available in `neko` and `neko-top` can
be found in @ref meshing.
In this tutorial we will be utilizing the `neko` utility `genmeshbox` in the 
`prepare.sh` script to generate \f$ \Omega \f$ from a cartesian mesh. The
key excerpts in this script read
```bash
# Handle options
Nx=30 && Ny=10 && Nz=10

...

echo "Generating mesh with dimensions: $Nx $Ny $Nz"
genmeshbox 0 6 0 2 0 2 $Nx $Ny $Nz .false. .false. .false.
```

indicating that a box mesh of dimensions \f$[0,6] \times [0,2] \times [0,2] \f$
will be created with \f$30 \times 10 \times 10 \f$ elements.
\note The `.false.` arguments refer to periodic boundary conditions, which in
`neko` are treated as topologically linking these boundary. In the current
problem there are no periodic boundary conditions, hence the `.false.` arguments

An important consideration when using `genmeshbox` is the boundary identifiers,

> which become relavent when prescribing boundary conditions. The convention is

Reviewer: ```suggestion
which become relevant when prescribing boundary conditions. The convention is
```

to allow for additional adjoint and optimization parameters. The `neko-top`
high level case structure reads
```
{
    "version": 1.0
    "case": {
        "numerics": {},
        "fluid": {},
        "scalar": {},
        "statistics": {},
        "simulation_components" : [],
        "point_zones" : [],
        "adjoint_fluid": {},
        "adjoint_scalar": {}
    },
    "optimization": {
        "domain": {},
        "design": {},
        "solver": {},
        "objectives": [],
        "constraints": []
    }
}

```
### Case
The top level of the case file allows the user to prescribe details on the
numerics and case set up.

```
{
    "version": 1.0,
    "case": {
        "mesh_file": "box.nmsh",
        "output_checkpoints": false,
        "output_at_end": false,
        "output_boundary": false,
        "end_time": 8.0,
        "timestep": 2e-4,
        "numerics": {
            "time_order": 1,
            "polynomial_order": 6,
            "dealias": true
        },
```
\note The default name when using `genmeshbox` is `"box.nmsh"`.

It is common practice in `neko-top` to disable the inbuilt `neko` file outputs,

> and instead rely on the output samplers provided by `neko-top` to aleviate

Reviewer: ```suggestion
and instead rely on the output samplers provided by `neko-top` to alleviate
```

> excessesive outputs when looping between the forward and adjoint runs.

Reviewer: ```suggestion
excessive outputs when looping between the forward and adjoint runs.
```

Here we have prescribed: 

> - a timestep \f$\Delta t = 2\times 10^{-4} \f$ (`"timestep": 2e-4`)

Reviewer: ```suggestion
- a time-step \f$\Delta t = 2\times 10^{-4} \f$ (`"timestep": 2e-4`)
```

- a first order time integration scheme (`"time_order": 1`)
- a polynomial order of 6 (`"polynomial_order": 6`)
- the use of overintegration when evaluating the convective terms (`"dealias": true`)

> \attention Despite intending to solve the steady Navier--Stokes equation,

Reviewer: ```suggestion
\attention Despite intending to solve the steady Navier-Stokes equation,
```

Currently an `end_time` must be prescribed in `neko`. More details regarding
steady state solutions can be found in [Simulation components](@ref simulation_components)  but there are
intentions to include steady state functionality directly within `neko`,
streamlining this process.

Instead, steady states are currently marched towards using the the steady
state simulation component
```
        "simulation_components": [
            {
                "type": "steady",
                "is_user": true,
                "tol": 1e-7,
                "compute_control": "tsteps",
                "compute_value": 1
            }
        ],
```

\attention Currently certain features of `neko` are not supported by `neko-top`

> due to a lack of adjoint support. Relavent to this section, it should be noted
> that varaible timestepping is not supported.

Reviewer: ```suggestion
due to a lack of adjoint support. Relevant to this section, it should be noted
that variable time-stepping is not supported.
```

                    "zone_indices": [2]
                }
            ],
            "output_control": "never",
        },
```

The key details which are prescribed are:
- a significantly low Reynolds number (`"Re": 2`)
- an initial \f$ \mathbf{u}|_{t=0} = \mathbf{0}\f$
```
            "initial_condition": {
                "type": "uniform",
                "value": [ 0.0, 0.0, 0.0 ]
            },
```
- no slip boundary conditions on \f$\Gamma_\text{wall}\f$
```
                    "type": "no_slip",
                    "zone_indices": [3, 4, 5, 6]
```
- an outflow boundary condition on \f$\Gamma_\text{out}\f$
```
                    "type": "outflow",
                    "zone_indices": [2]

```
- a user defined inflow condition on \f$\Gamma_\text{in}\f$
```
                    "type": "user_velocity_pointwise",
                    "zone_indices": [1]
```

\note The velocity and pressure solvers have been set relatively conventionally,
however the interested reader can find more information 
[here](https://neko.cfd/docs/release/dd/d33/case-file.html#autotoc_md39). 

\attention
It should be noted that some features in `neko` are not supported in `neko-top`

> due to a lack of adjoint support. Relavent to this section, it should be noted

Reviewer: ```suggestion
due to a lack of adjoint support. Relevant to this section, it should be noted
```

that the use of projections or gradient jump penalty are not supported.

To prescribe user defined conditions, such as the condition on \f$\Gamma_\text{in}\f$
users can provide additional fortran code. In this tutorial we will provide an
additional `user.f90`. Further details on the `neko` user interface can be
found [here](https://neko.cfd/docs/release/d6/def/user-file.html).

> In this tutorial we wish to prescribe a parabaloid as an inflow condition, this

Reviewer: ```suggestion
In this tutorial we wish to prescribe a paraboloid as an inflow condition, this
```

                    "flux": 0,
                    "zone_indices": [2, 3, 4, 5, 6]
                },
            ],
        },
```
The key details which are prescribed are:
- a significantly high Peclet number (`"Pe": 5000.0`)
- a user defined initial condition
```
            "initial_condition": {
                "type": "user"
            },
```
- Neuman conditions on \f$ \Gamma_\text{out} \cup \Gamma_\text{wall} \f$
```
                    "type": "neumann",
                    "flux": 0,
                    "zone_indices": [2, 3, 4, 5, 6]
```
- a user defined scalar inflow condition on \f$ \Gamma_\text{in}\f$
```
                    "type": "user_pointwise",
                    "zone_indices": [1]
```

The user defined scalar inflow condition and scalar initial condition are 
prescribed with `user%scalar_user_bc` and `user%scalar_user_ic`. Here we wish
to prescribe a separation between the two scalar species where 
\f$\phi = 0\f$ represents one species and \f$\phi = 1\f$ represents the other.
It is well understood that spectral methods struggle to represent discontinuities,
hence, a sigmoid function is use to provide a sufficient amount of smoothing

> accross the inteface of the two species. The completed `user.f90` should now

Reviewer: ```suggestion
across the interface of the two species. The completed `user.f90` should now
```

            ],
        },
```
\note An important feature to be aware of in the `neko-top` adjoint system is that
unless a parameter is specified, the adjoint will adopt the same parameters
as their primal counterparts. For instance, by not specifying a solver type
for the adjoint fluid, `neko-top` will use the same solver types used in the
fluid.

The key details which are prescribed are:
- \f$ \phi^\dagger = 0\f$ on \f$ \Gamma_\text{in} \f$
```
                    "type": "dirichlet",
                    "value": 0,
                    "zone_indices": [1]
```
- \f$ \nabla  \phi^\dagger \cdot \mathbf{n} = 0\f$ on 
\f$ \Gamma_\text{out} \cup \Gamma_\text{wall} \f$
```
                    "type": "neumann",
                    "flux": 0,
                    "zone_indices": [2, 3, 4, 5, 6]
```
- \f$ \mathbf{u}^\dagger = \mathbf{0}\f$ on 
\f$ \Gamma_\text{in} \cup \Gamma_\text{wall} \f$
```
                    "type": "no_slip",
                    "zone_indices": [1, 3, 4, 5, 6]
```
- an outflow velocity boundary condition on \f$ \Gamma_\text{out} \f$
```
                     "type": "outflow",
                     "zone_indices": [2]
```

## Domain specification {#mixer_domains}

> Domains can be specified through the `neko` functionility `point_zones`, the
> reader is reffered [here](https://neko.cfd/docs/release/da/dd0/point-zones.html)
> for extensive documentation on point zones.

Reviewer: ```suggestion
Domains can be specified through the `neko` functionality `point_zones`, the
reader is referred [here](https://neko.cfd/docs/release/da/dd0/point-zones.html)
for extensive documentation on point zones.
```

                   -1.0,
                    3.0,
                ],
            },
            {
                "name": "objective_domain",
                "geometry": "box",
                "x_bounds": [
                    5.0,
                    6.0,
                ],
                "y_bounds": [
                    -1.0,
                     3.0,
                ],
                "z_bounds": [
                   -1.0,
                    3.0,
                ],
            },
        ]
    },
```

Assigning \f$\Omega_\text{opt}\f$ to the optimization domain is performed in
the `optimization` section of the case file
```
    "optimization": {
        "domain": {
            "type": "point_zone",
            "zone_name": "optimization_domain"
            },
```
## Objectives and constraint {#mixer_objectives}
Various objectives and constraints can be provided as lists in the `objectives`
and `constraints` subsection of `optimization`. A full list of the available
objectives and constraints can be found in [Objectives and constraints](@ref objectives_and_constraints).

\attention In the future we wish to also include a tutorial teaching users how

> to implement their own custome objectives and constraints, however this is yet

Reviewer: ```suggestion
to implement their own custom objectives and constraints, however this is yet
```

concentration to \f$ \phi = 0.5\f$ (`"phi_ref": 0.5`) and assert the area of
interest is in \f$ \Omega_{mix}\f$ (`"mask_name": "objective_domain"`).

## Mapping cascade {#mixer_mapping}
We saw in [Defining the optimization problem](@ref mixer_optimization_problem) that the design indicator field
\f$\rho\f$ must be mapped to the Brinkman amplitude \f$\chi\f$. This is achieved
through the use of the mapping cascade accessible in `optimization.design.mapping`.
A detailed description of the mapping cascade and available mappings in `neko-top`
can be found in [Mapping cascade](@ref mapping_cascade).

In this tutorial we use the composite mapping 
\f$\rho \mapsto \tilde{\rho} \mapsto \chi\f$ where
- \f$\rho \mapsto \tilde{\rho} \f$ is a spatial filter
-\f$\tilde{\rho} \mapsto \chi\f$ is a RAMP mapping

which can be prescribed with
```
        "design": {
            "type": "brinkman",
            "mapping" : [
                {
                    "type": "PDE_filter",
                    "r": 0.01
                },
                {
                    "type": "RAMP",
                    "f_max": 1000
                }
            ],
```

> \note It is important to note that the order in which the mappings occure in

Reviewer: ```suggestion
\note It is important to note that the order in which the mappings occur in
```

the case file is the order in which they will be executed.

Here we have prescribed a spatial filter with radius 1 followed by a RAMP mapping
with maximum Brinkman amplitude 1000.

The mapping cascade takes care of both mapping the design forward and performing
the chain rule backward, i.e. 
\f[
    \frac{\partial F}{\partial \chi}
    \mapsto \frac{\partial F}{\partial \tilde{\rho}}
    \mapsto \frac{\partial F}{\partial \rho}
\f]

## Optimization parameters {#mixer_MMA}
The optimization algorithm used to solve optimization problem is prescribed
in the `"solver"` section of the case file.
Currently on the MMA algorithm is implemented and more details can be found
in [The Method of Moving Asymptotes (MMA)](@ref MMA).

In this tutorial we we select the MMA algorithm by prescribing
```
        "solver": {
            "type": "mma",
            "max_iterations": 100,
            "tolerance": 1.0e-3,
            "mma": {
                "m": 0,
                "scale": 1000.0,
                "auto_scale": false,
                "xmin": 0.0,
                "xmax": 1.0
            }
        },
```

> The key paramters selected are

Reviewer: ```suggestion
The key parameters selected are
```

- a maximum of 100 optimization iterations (`"max_iter": 100`)

> - a convergence criterior of 1.0e-3 (`"tolerance": 1.0e-3`)

Reviewer: ```suggestion
- a convergence criterion of 1.0e-3 (`"tolerance": 1.0e-3`)
```

- no constraints (`"m": 0`)
- a scaling factor of 1000 for the objective function (`"scale": 1000.0`)
- assert that \f$ \rho \in [0,1] \f$
```
                "xmin": 0.0,
                "xmax": 1.0
```

Scaling factors can be adjusted to influence convergence rates of the optimization
problem.

## Neko-top drivers {#mixer_driver}
`neko-top` exists as a library of tools to allow users to perform optimization
in conjunction with the `neko` library. As such, the user has the freedom to
create custom drivers to perform specific tasks. Further details regarding drivers
can be found in [Advanced example setups](@ref examples-advanced).

This functionality can be achieved by providing a `driver.f90` file in the

> example folder along with a `CMakelist.txt`.

Reviewer: ```suggestion
example folder along with a `CMakeLists.txt`.
```

> For this tutorial, we provide the following `CMakelist.txt`

Reviewer: ```suggestion
For this tutorial, we provide the following `CMakeLists.txt`
```

```CMake
# Simple setup to compile the current example.
set(DRIVER_TYPE "custom")
set(DRIVER_SOURCE ${CMAKE_CURRENT_SOURCE_DIR}/driver.f90)
set(EXTRA_SOURCES
    ${CMAKE_CURRENT_SOURCE_DIR}/user.f90
)

build_example(${DRIVER_TYPE})
```

This script performs the following task:
- allows `neko-top` to use a custom driver (`set(DRIVER_TYPE "custom")`)
- indicates to `neko-top` to use the `driver.f90` driver (`set(DRIVER_SOURCE ${CMAKE_CURRENT_SOURCE_DIR}/driver.f90)`)
- includes the additional `user.f90` file used to prescribe user boundary conditions
and initial conditions.
```
set(EXTRA_SOURCES
    ${CMAKE_CURRENT_SOURCE_DIR}/user.f90
)
```

`neko-top` distinguishes between four key objects:
- [design](@ref design). The design space over which one optimizes.
- [problem](@ref problem). The definition of the optimization problem being solved.

> - [optimizer](@ref optimizer). The optimization algoirthm used to solve the optimization

Reviewer: ```suggestion
- [optimizer](@ref optimizer). The optimization algorithm used to solve the optimization
```

problem.
- [simulation](@ref simulation). Specifically for problems involving fluid mechanics, a
simulation can allow for interfaces with `neko` to perform forward simulation
and additional `neko-top` libraries to perform adjoint sensitivity analysis.

> Further details regarding these componenents can be found in [Code structure ](@ref code-structure).

Reviewer: ```suggestion
Further details regarding these components can be found in [Code structure ](@ref code-structure).
```
